### DISCUSSION TITLE
[closed][x][manuscript][blueprint] sample:sample_type:term_iri

### DISCUSSION THREAD

**Discussion by dzerbino** [fairtracks/fairtracks_standard#96] [Team Posts]
Sometimes people like to have a set of IRIs, eg.:
http://purl.obolibrary.org/obo/CL_0000096 AND http://purl.obolibrary.org/obo/UBERON_0013756

How to deal with those?

**Comment by sveinugu** [comment]
Don't know. Array is a possibility, but creates issues with search. We are very happy that we got rid of them in version 2!..

**Comment by sveinugu** [comment]
One idea, given that there is a hierarchy (which there is in your example), e.g. venous blood -> mature neutrophil, is to have a primary and a secondary sample_type, somehow properly named. Will that solve most of the cases?

**Comment by dzerbino** [comment]
I don't think so. If one of the terms subsumes the other, then it is redundant. Normally the two terms would not be related.

**Comment by sveinugu** [comment]
@dtitov Does arrays still work in TrackFind search, or did you remove the support?

**Comment by dtitov** [comment]
@sveinugu, I removed it since I was expecting that the new version of schema will never have arrays (or at least the array-type attributes will not be used in the search-queries).

**Comment by sveinugu** [comment]
Would it be difficult to reimplement it for search? Also, when generating all values, it should understand arrays. A possible goal is that if one sample has e.g.:
```
"sample_type": [
    {
        "term_iri": "http://purl.obolibrary.org/obo/CL_0000096",
        "term_value": "mature neutrophil"
    },{
        "term_iri": "http://purl.obolibrary.org/obo/UBERON_0013756",
        "term_value": "venous blood"
    }
]
```
than the term_iri value list should contain both values, likewise for the term_values. The values should be possible to use in search. Also, if one searches for both these values, the sample record should only appear only once.

Would this be possible, or too complicated?

**Comment by sveinugu** [comment]
Ah, then we will also have the issue of the GSuite conversion.. :(

**Comment by dtitov** [comment]
1. I didn't really like how it was implemented before. Now our queries are SQL-compliant and I find it to be a benefit. Before it was a "hack" when I performed custom processing of asterisk symbol (`*`), which was not a real SQL.
2. Probably it is still possible to implement it, but it will be more complicated than before, because that asterisk-hack was based on introducing custom joins. And now we already have joins (for categories). So I'll have to put joins on top of joins 😕
3. And, yes, it seems that we will still have a problem with conversion to GSuite...

**Comment by dzerbino** [comment]
OK, I'm getting the impression that this will be difficult in a first pass. I see two way forwards in the short term:
- the spec requires only one term, and I drop one at random
- the spec accepts a list of terms, and you drop one term at random. 

It's not as bad as it sounds, since there are few instances of such double-barrelled terms.

**Comment by sveinugu** [comment]
Opening again, due to https://github.com/orgs/fairtracks/teams/fairtracks-developers/discussions/28. @dzerbino If the tissue_type terms are removed, are there any double-barrelled terms left in sample_type?

**Comment by dzerbino** [comment]
It would, and that's what I did in the latest JSON.

**Comment by sveinugu** [comment]
As far as I can see, the tissue_types have not been removed from the sample_type in the latest json. You can check for yourself in TrackFind: https://trackfind-dev.gtrack.no/ E.g. «Bone marrow” are in both lists.

**Comment by dzerbino** [comment]
Yes, these samples only had one ontology term, and it was UBERON, so it remained for those samples.

**Comment by sveinugu** [comment]
@dzerbino Yes, I see. I agree the best solution is to keep them there, then, as the tissue terms at least narrow down the cell types (the ones we are talking about is "bone marrow", "venous blood", "capillary blood", and "lymph node"). Looking at the EFO, I see no better solution. 

Edit: moved the tissue_type comment to https://github.com/orgs/fairtracks/teams/fairtracks-developers/discussions/28.

@dzerbino So if I understand you correctly, there are still samples that are annotated with more than one sample type term even when you have removed the tissue_type terms? Are there many of those? Could you provide some examples?

**Comment by sveinugu** [comment]
The main question being whether we should (in the future) implement sample_type as an array or not. I'm not really at ease with the current solution of just choosing one (the first?)

**Comment by dzerbino** [comment]
I believe the long term solution is absolutely to encode them as arrays. 

Currently we only have one term for convenience.

**Comment by sveinugu** [comment]
It seems to me that all the pairs in BLUEPRINT is a combination of cell type and tissue type (CL and UBERON). So no need for an array, as far as I can understand.

**Comment by dzerbino** [comment]
In the short term yes, but, to repeat, future proofing would require us to change this single ontology term to a a list of ontology terms.

**Comment by sveinugu** [comment]
@dzerbino @jmfernandez 
Should be enough possibilities in this suggestion, which increases preciseness, while opening up for several subtypes of sample classification: https://github.com/fairtracks/fairtracks_standard/pull/20. I am very much against supporting arrays. If there is new types of sample data that is required, we can add new subproperties for this.

**Comment by sveinugu** [comment]
@dzerbino Can I close this?

**Comment by dzerbino** [comment]
Hello @sveinugu, why are you against supporting arrays? Updating the spec by adding subproperties does not sound sustainable to me.

**Comment by sveinugu** [comment]
@dzerbino I am not against supporting arrays as such, just that I have over the years never seen any track metadata with arrays in sample_type fields that is not due to unclear metadata. Also in the case of BLUEPRINT, the reason for the arrays were that different fields (tissue type and cell type) were merged. Provided with metadata that really needs arrays (e.g. a sample of multiple specific cell types), I would be happy to add them. For downstream analysis one would typically create tables or heat maps where one column/dimension would be e.g. cell type. One would not want to have the same track used more than one place in such results, so one would either have to specify one cell type as the most important, or one would need to include all cell types as a merged cell type category specific for the combination.  And in the latter case, one would need to make sure that the order of the cell_types would always be the same, e.g. alphabetically, so that it can be used as a category. It just opens up a world of hurt downstream, and I don’t want to go there unless absolutely necessary. IHEC does not support arrays for such metadata fields, as far as I know.

**Comment by dzerbino** [comment]
Sorry.... https://github.com/IHEC/ihec-ecosystems/issues/47

**Comment by sveinugu** [comment]
@dzerbino So this is the same BLUEPRINT use case as you have been pushing here, and I argue that it is wrong. Venous blood is the tissue type, and it fits under organism_part in the version of EFO defined in the current Schema. So I would say your use case for adding support for arrays in IHEC is equally wrong also there. The solution in IHEC would be to add also tissue_type to the fields for Primary Cell and Primary Cell Culture, as just the cell type is too little information. This is what I did in my pull request (see ‘organism_part’ description in https://github.com/fairtracks/fairtracks_standard/pull/20/files). 

Btw: EFO just removed over 10000 UBERON terms, including «venous blood». However, this does not invalidate the argument above. I added another discussion point on the EFO issue, though: https://github.com/orgs/fairtracks/teams/fairtracks-developers/discussions/47

**Comment by sveinugu** [comment]
@dzerbino As I think I mentioned above, I could be convinced by other use cases, but my hunch is that such cases are too rare to complicate things for.

**Comment by sveinugu** [comment]
@dzerbino ....aaaand another argument: opening up for too many possibilites for the curators increases the likelihood of low quality metadata.

**Comment by sveinugu** [comment]
@dzerbino Still waiting for some counterarguments. For the future, I suppose, as you did accept my PR.

**Comment by sveinugu** [comment]
Linked to this in https://github.com/orgs/fairtracks/teams/fairtracks-developers/discussions/41. Closing
